src/type/validate.js

```diff
@@ -593,6 +759,143 @@ function createInputObjectCircularRefsValidator(
+
+      // Only input object type fields can result in a cycle.
+      if (!isInputObjectType(fieldType)) {
+        continue;
+      }
+
+      // If the provided value has this field defined, recursively check it for
+      // cycles instead of checking the field's default value.
+      if (hasOwnProperty(defaultValue, field.name)) {
+        detectValueDefaultValueCycle(fieldType, defaultValue[field.name]);
+        continue;
+      }
+
+      detectFieldDefaultValueCycle(inputObj, field, fieldType);
+    }
+  }
+
+  function detectLiteralDefaultValueCycle(
+    inputObj: GraphQLInputObjectType,
+    defaultValue: ConstValueNode,
+  ): void {
+    // If the value is a List, recursively check each entry for a cycle.
+    // Otherwise, only object values can contain a cycle.
+    if (defaultValue.kind === Kind.LIST) {
+      for (const itemLiteral of defaultValue.values) {
+        detectLiteralDefaultValueCycle(inputObj, itemLiteral);
+      }
+      return;
+    } else if (defaultValue.kind !== Kind.OBJECT) {
+      return;
+    }
+
+    // Check each defined field for a cycle.
+    const fieldNodes = keyMap(defaultValue.fields, (field) => field.name.value);
+    for (const field of Object.values(inputObj.getFields())) {
+      const fieldType = getNamedType(field.type);
+
+      // Only input object type fields can result in a cycle.
+      if (!isInputObjectType(fieldType)) {
+        continue;
+      }
+
+      // If the provided value has this field defined, recursively check it for
+      // cycles instead of checking the field's default value.
+      if (hasOwnProperty(fieldNodes, field.name)) {
+        detectLiteralDefaultValueCycle(fieldType, fieldNodes[field.name].value);
+        continue;
+      }
+
+      detectFieldDefaultValueCycle(inputObj, field, fieldType);
+    }
+  }
+
+  function detectFieldDefaultValueCycle(
+    inputObj: GraphQLInputObjectType,
+    field: GraphQLInputField,
+    fieldType: GraphQLInputObjectType,
+  ): void {
+    // Only a field with a default value can result in a cycle
+    const defaultValue = field.defaultValue;
+    if (!defaultValue) {
+      return;
+    }
+
+    const fieldCoordinate = `${inputObj.name}.${field.name}`;
+
+    // Check to see if there is cycle.
+    const cycleIndex = fieldPathIndex[fieldCoordinate];
+    if (cycleIndex >= 0) {
+      const cyclePath = fieldPath.slice(cycleIndex);
+      const pathStr = cyclePath.map(([coordinate]) => coordinate).join(', ');
+      context.reportError(
+        `Cannot reference Input Object field ${fieldCoordinate} within itself through a series of default values: ${pathStr}.`,
+        cyclePath.map(([, node]) => node),
+      );
+      return;
+    }
+
+    // Recurse into this field's default value once, tracking the path.
+    if (!visitedFields[fieldCoordinate]) {
+      visitedFields[fieldCoordinate] = true;
+      fieldPathIndex[fieldCoordinate] = fieldPath.length;
+      fieldPath.push([fieldCoordinate, field.astNode?.defaultValue]);
```

Comment:
I've found that this pattern is risky from a maintenance POV because the two lines of code can get split over time which may at some point cause the index to diverge by accident. I use the fact that `push` returns the new length of the array to make it one cohesive operation:

```suggestion
      fieldPathIndex[fieldCoordinate] = fieldPath.push([fieldCoordinate, field.astNode?.defaultValue]) - 1;
```

Feel free to completely ignore this suggestion, it's just personal preference.